src/modules/tests/tests.service.ts

```diff
@@ -419,48 +419,284 @@ export class TestsService {
+        canResume: false, // No attempts to resume
+        attemptGrading: test.attemptsGrading,
+        gradedAttempt: null, // No graded attempt yet
+        lastAttempt: null, // No last attempt
+        showCorrectAnswers: test.showCorrectAnswer,
+      };
+    }
+
+    // Check if there are any attempts under review (lightweight query)
+    const hasPendingReview = await this.hasPendingReview(testId, userId, authContext);
+
+    // Get graded attempt based on grading method (optimized query)
+    let finalScore: number = 0;
+    let finalResult: string | null = null;
+    let finalAttemptId: string | null = null;
+    let gradedAttemptData: TestAttempt | null = null;
+
+    switch (test.attemptsGrading) {
+      case AttemptsGradeMethod.FIRST_ATTEMPT: {
+        gradedAttemptData = await this.getFirstAttempt(testId, userId, authContext);
+        if (gradedAttemptData) {
+          finalScore = gradedAttemptData.score || 0;
+          finalResult = gradedAttemptData.result || null;
+          finalAttemptId = gradedAttemptData.attemptId;
+        }
+        break;
+      }
+
+      case AttemptsGradeMethod.LAST_ATTEMPT: {
+        gradedAttemptData = await this.getLastAttempt(testId, userId, authContext);
+        if (gradedAttemptData) {
+          finalScore = gradedAttemptData.score || 0;
+          finalResult = gradedAttemptData.result || null;
+          finalAttemptId = gradedAttemptData.attemptId;
+        }
+        break;
+      }
+
+      case AttemptsGradeMethod.HIGHEST: {
+        gradedAttemptData = await this.getHighestAttempt(testId, userId, authContext);
+        if (gradedAttemptData) {
+          finalScore = gradedAttemptData.score || 0;
+          finalResult = gradedAttemptData.result || null;
+          finalAttemptId = gradedAttemptData.attemptId;
+        }
+        break;
+      }
+
+      case AttemptsGradeMethod.AVERAGE: {
+        const averageData = await this.getAverageScore(testId, userId, authContext);
+        if (averageData) {
+          finalScore = averageData.averageScore;
+          finalResult = finalScore >= test.passingMarks ? 'P' : 'F'; // PASS or FAIL
+          
+          // Get the last submitted attempt for attemptId reference
+          const lastSubmittedAttempt = await this.getLastAttempt(testId, userId, authContext);
+          if (lastSubmittedAttempt) {
+            finalAttemptId = lastSubmittedAttempt.attemptId;
+            gradedAttemptData = lastSubmittedAttempt; // Use for graded attempt object
+          }
+        }
+        break;
+      }
+    }
+
+    // If there are any attempts under review, set finalResult to null
+    if (hasPendingReview) {
+      finalResult = null;
+    }
+
+    // Check if user can attempt (hasn't reached max attempts)
+    const canAttempt = totalAttempts < maxAttempts;
+
+    // Get last attempt for resume check (lightweight query)
+    const lastAttempt = await this.getLastAttemptForResume(testId, userId, authContext);
```

Comment:
@vinayak-patil what is difference between getLastAttempt and getLastAttemptForResume functions?

Reply:
@vinayak-patil getLastAttempt change this to lastCompletedAttempt. Also while test is created, the option should correctly mention this.